util/chplenv/printchplbuilds.py

```diff
@@ -0,0 +1,449 @@
+        if options.csh:
+            parser.error("--check cannot be used with --csh.")
+        if options.bash:
+            parser.error("--check cannot be used with --bash.")
+
+
+    # Note: "allvars" is a list of all possible variables in the order in
+    # which they are displayed by printchplenv so we can print them in the
+    # same order. "usedvars" is a list of the variables actually used in the
+    # configurations, in the same order. "used" is a set of used variables to
+    # make it easy to determine whether a variable has been used or
+    # not. "usedvars" is created from "allvars" and "used".
+
+    # get the names of all variables
+    printchplenv.compute_all_values()
+    allvars = GetConfig(True).keys()
+    used = set()
+
+    # get the current configuration
+    current = GetConfig()
+    for var in current:
+        used.add(var)
+
+    if options.path is not None:
+        builds = [Parse(options.path.split(os.sep))]
+    else:
+        # gather the paths for all builds
+        home = os.getenv('CHPL_HOME')
+        if home is None:
+            print("Error: CHPL_HOME is not set", file=sys.stderr)
+            sys.exit(1)
+
+        base = os.path.join(home, "lib")
+        # "targets" is a list of all paths that contain a "main.o" file
+        targets = []
+        for root, dirs, files in os.walk(base):
+            for target in filter(lambda x: x == "main.o", files):
+                targets.append(os.path.join(root, target))
+
+        if options.oldest:
+            targets.sort(key= lambda x: os.stat(x)[stat.ST_MTIME])
+
+        if options.newest:
+            targets.sort(key= lambda x: os.stat(x)[stat.ST_MTIME], reverse=True)
+
+        # Convert the targets into a list of paths in which each path is a list of
+        # components, excluding the final "main.o"
+        paths = []
+        for target in targets:
+            relpath = os.path.relpath(target, base)
+            paths.append(relpath.split(os.sep)[:-1])
+
+        builds = []
+        for (target, path) in zip(targets, paths):
+            build = Parse(path)
+            build['MTIME'] = os.stat(target)[stat.ST_MTIME]
+            builds.append(build)
+
+    # make a list of variables that were actually used by the configurations
+    usedvars = [x for x in allvars if x in used]
+
+    if options.summary:
+        # summarize which variables differ between the configs and their values
+        diffvars = set()
+        for i in range(len(builds)):
+            diffvars |= set(FindDiffVars(usedvars, current, builds[i]))
+        values = {}
+        for config in [current] + builds:
+            for var in usedvars:
+                if var in diffvars:
+                    value = config.get(var, 'NA')
+                    if var not in values.keys():
+                        values[var] = set([value])
+                    else:
+                        values[var].add(value)
```

Comment:
values can be a `defaultdict(set)` here and then do `values[var].add(value)` for a minor cleanup